api/src/extensions/manager.ts

```diff
@@ -322,17 +323,30 @@ export class ExtensionManager {
 			}
 		});
 
-		this.watcher = chokidar.watch(
-			[path.resolve('package.json'), path.posix.join(extensionDirUrl, '*', 'package.json'), ...localExtensionUrls],
-			{
-				ignoreInitial: true,
-			},
-		);
+		const rootPackageJson = path.resolve(env['PACKAGE_FILE_LOCATION'] as string, 'package.json');
+		const localExtensions = path.posix.join(extensionsDir, '*', 'package.json');
+
+		this.watcher = chokidar.watch([rootPackageJson, localExtensions, ...nestedExtensions], {
+			ignoreInitial: true,
+			// dotdirs are watched by default and frequently found in 'node_modules'
+			ignored: `${extensionsDir}/**/node_modules/**`,
+			// on macOS dotdirs in linked extensions are watched too
+			followSymlinks: os.platform() === 'darwin' ? false : true,
+		});
 
 		this.watcher
-			.on('add', () => this.reload())
-			.on('change', () => this.reload())
-			.on('unlink', () => this.reload());
+			.on(
+				'add',
+				debounce(() => this.reload(), 500),
+			)
+			.on(
+				'change',
+				debounce(() => this.reload(), 650),
+			)
+			.on(
+				'unlink',
+				debounce(() => this.reload(), 2000),
+			);
```

Comment:
I'd rather not make this configurable. I don't see a real reason for this.
The debouncing is intended as an optimization, with values that will cover most cases.
However, we can never cover all cases (for example edge-cases like in #20440).
Even if we make it configurable, it doesn't really make sense to set the values higher, because then it takes too long for single file changes.
On the other side, IMO the values are small enough not to impair everyday use.